base/gfspapp/download_server.go

```diff
@@ -66,13 +62,55 @@ func (g *GfSpBaseApp) OnDownloadObjectTask(
 	return data, nil
 }
 
-func (g *GfSpBaseApp) GfSpGetChallengeInfo(
-	ctx context.Context,
-	req *gfspserver.GfSpGetChallengeInfoRequest) (
+func (g *GfSpBaseApp) GfSpDownloadPiece(ctx context.Context, req *gfspserver.GfSpDownloadPieceRequest) (
+	*gfspserver.GfSpDownloadPieceResponse, error) {
+	task := req.GetDownloadPieceTask()
+	if task == nil {
+		log.Error("failed to download piece due to task pointer dangling")
+		return &gfspserver.GfSpDownloadPieceResponse{Err: ErrDownloadTaskDangling}, nil
+	}
+	ctx = log.WithValue(ctx, log.CtxKeyTask, task.Key().String())
+	span, err := g.downloader.ReserveResource(ctx, task.EstimateLimit().ScopeStat())
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to reserve download piece resource", "error", err)
+		return &gfspserver.GfSpDownloadPieceResponse{Err: ErrDownloadExhaustResource}, nil
+	}
+	defer span.Done()
+	metrics.DownloadPieceSizeHistogram.WithLabelValues(
+		g.downloader.Name()).Observe(float64(task.GetSize()))
+	data, err := g.OnDownloadPieceTask(ctx, task)
+	log.CtxDebugw(ctx, "finished to download piece", "len", len(data), "error", err)
+	return &gfspserver.GfSpDownloadPieceResponse{
+		Err:  gfsperrors.MakeGfSpError(err),
+		Data: data}, nil
+}
+
+func (g *GfSpBaseApp) OnDownloadPieceTask(ctx context.Context, task task.DownloadPieceTask) (
```

Comment:
ditto. I found that task var name is too many, may change the imported package name.

Reply:
Updated.